src/libdredd/src/mutation_replace_binary_operator.cc

```diff
@@ -1024,34 +1029,108 @@ MutationReplaceBinaryOperator::ClangOperatorKindToProtobufOperatorKind(
+      return replacement_operator != clang::BO_LE &&
+             replacement_operator != clang::BO_NE;
     case clang::BO_LE:
-      return operator_kind != clang::BO_LT && operator_kind != clang::BO_EQ;
+      return replacement_operator != clang::BO_LT &&
+             replacement_operator != clang::BO_EQ;
     case clang::BO_EQ:
-      return operator_kind != clang::BO_LE && operator_kind != clang::BO_GE;
+      return replacement_operator != clang::BO_LE &&
+             replacement_operator != clang::BO_GE;
     case clang::BO_NE:
-      return operator_kind != clang::BO_LT && operator_kind != clang::BO_GT;
+      return replacement_operator != clang::BO_LT &&
+             replacement_operator != clang::BO_GT;
     default:
       return false;
   }
 }
 
+bool MutationReplaceBinaryOperator::IsRedundantReplacementForUnsignedComparison(
+    clang::BinaryOperatorKind replacement_operator,
+    const clang::ASTContext& ast_context) const {
+  if (!binary_operator_->getLHS()->getType()->isUnsignedIntegerType() ||
+      !binary_operator_->getRHS()->getType()->isUnsignedIntegerType()) {
+    return false;
+  }
+  if (MutationReplaceExpr::ExprIsEquivalentToInt(*binary_operator_->getLHS(), 0,
+                                                 ast_context)) {
+    // LHS is 0
+    if (replacement_operator == clang::BO_GT ||
+        replacement_operator == clang::BO_LE) {
+      // We are considering mutating an expression of the form "0 op a" to
+      // one of:
+      //
+      // - "0 > a"
+      // - "0 <= a"
+      //
+      // These are subsumed by replacing the overall expression with "false"
+      // and "true", respectively.
+      return true;
+    }
+    if (std::set({binary_operator_->getOpcode(), replacement_operator}) ==
+        std::set<clang::BinaryOperatorKind>({clang::BO_EQ, clang::BO_GE})) {
+      // "0 == a" is equivalent to "0 >= a"
+      return true;
+    }
+    if (std::set({binary_operator_->getOpcode(), replacement_operator}) ==
+        std::set<clang::BinaryOperatorKind>({clang::BO_NE, clang::BO_LT})) {
+      // "0 != a" equivalent to "0 < a"
+      return true;
+    }
+  }
+  if (MutationReplaceExpr::ExprIsEquivalentToInt(*binary_operator_->getRHS(), 0,
+                                                 ast_context)) {
+    // RHS is 0
+    if (replacement_operator == clang::BO_LT ||
+        replacement_operator == clang::BO_GE) {
+      // We are considering mutating an expression of the form "a op 0" to
+      // one of:
+      //
+      // - "a < 0"
+      // - "a >= 0"
+      //
+      // These are subsumed by replacing the overall expression with "false"
+      // and "true", respectively.
+      return true;
+    }
+    if (binary_operator_->getOpcode() == clang::BO_EQ &&
+        replacement_operator == clang::BO_LE) {
+      // "a == 0" is equivalent to "a <= 0"
+      return true;
+    }
+    if (binary_operator_->getOpcode() == clang::BO_NE &&
+        replacement_operator == clang::BO_GT) {
```

Comment:
Above you used set equality but here used two equality checks, is there any reason?